raft-otel/README.md

```diff
@@ -0,0 +1,112 @@
+
+You can either 
+ 1) try to write your own RAFT implementation, building up the functionality in the stages described by Bendersky, or
+ 2) use Bendersky's code, after having thoroughly read it and understood it
+ 3) write your own implementation, using Bendersky's as a reference if you get stuck
+
+Reading code written by others is a useful skill to have, so if you opt to create your own implementation, you should review Bendersky's code.
+Does it differ from yours in any significant respect?
+
+By the end of Part 1 we should have a running RAFT cluster with 5 instances. 
+We may choose to run our RAFT cluster locally using `docker-compose`, `minikube`, or any other appropriate tool.
+
+### Part 2: Add distributed tracing
+
+In this section we will add distributed tracing support to the RAFT implementation from Part 1. 
+We will use [Honeycomb](https://www.honeycomb.io/), a SaaS distributed tracing provider. 
+
+Honeycomb provides a useful guide for their own 
+[OpenTelemetry Distribution for Go](https://docs.honeycomb.io/getting-data-in/opentelemetry/go-distro/).
+
+Add tracing to all parts of your application which you might wish to trace. Consider what operations are 
+interesting - in general, anything that may feasibly take a long time, such as a RPC, writing to storage, or
+taking a lock may be a candidate for a span. 
+
+Run your system and view traces in Honeycomb. Run through the [Honeycomb sandbox tour](https://play.honeycomb.io/sandbox/environments/analyze-debug-tour)
+and then explore your own data in the same way. 
+
+Do distributed operations (such as leader elections) surface in Honeycomb as coherent traces?
+
+Create a [Board](https://docs.honeycomb.io/working-with-your-data/boards/) in honeycomb with some useful visualisations.
+
+### Part 3: Debugging latency and failures using distributed tracing
+
+Bendersky's implementation of RAFT allows you to simulate unreliable RPCs. Enable this (if you wrote your own implementation then you will 
+need to add this kind of chaos capability as a feature first). 
+Now, use Honeycomb to observe the dropped RPCs and delays that 
+the simulation injects. Did these show up on your board from Part 2?
+
+Can you add further kinds of chaos? What about storage failures? 
+
+### Part 4: Reducing costs of distributed tracing
+
+In this exercise we are running a small system and using Honeycomb's free tier, so cost is not a consideration.
+However, in real production systems, distributed tracing can create a large volume of traces and spans. This can be 
+costly in terms of network, storage, or SaaS bills. 
+
+For this reason, many distributed tracing users use sampling or ratelimiting to control the number of traces that are
+collected. Read about [OTel Sampling and Ratelimiting](https://uptrace.dev/opentelemetry/sampling.html).
+
+Modify your solution to support sampling a specific percentage of requests, and to limit the total number of traces sent to no more than 100 per minute.
```

Comment:
```suggestion
Modify your solution to support sampling a specific percentage of requests, and to limit the total number of traces sent to no more than 100 per minute. (Hint: You typically want to get all or no spans for a whole trace, rather than dropping spans independently of the trace they're in).

Consider also whether some requests may be more important to trace than others. What may make a request more or less interesting than others?
```